src/embedded_flash/upgrade_storage.rs

```diff
@@ -0,0 +1,30 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use persistent_store::StorageResult;
+
+/// Accessors to storage locations used for upgrading from a CTAP command.
+pub trait UpgradeStorage {
+    /// Reads a slice of the partition, if within bounds.
+    fn read_partition(&self, offset: usize, length: usize) -> StorageResult<&[u8]>;
```

Comment:
Is the length of the partition a constant? If yes let's document it. Otherwise let's add a function to know it.

```rust
fn partition_length(&self) -> usize;
```

Reply:
I added `partition_length`, and no `partition_offset` or anything like that, since the indexing is now actually offset by the partition start.